# ==============================================================================
# FIXING COMMIT DIFF
# ==============================================================================
# File: accessible/mac/HyperTextAccessibleWrap.mm
# Commit: 582424635640
# Full Hash: 5824246356401f8bc1f978dc1bc1bfa4b11586ae
# Author: Eitan Isaacson <eitan@monotonous.org>
# Date: 2020-12-01 04:19:24
# Description:
#   Bug 1679075 - Don't go past empty input field in HyperTextIterator::Next. r=morgan
#   
#   The normalize function was putting the current container/offset past the end boundary. This caused overruns with getting the text for the range of the input object, and some unexpected invalid ranges.
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D98236
# ==============================================================================

diff -r b12f5962b9d2 -r 582424635640 accessible/mac/HyperTextAccessibleWrap.mm
--- a/accessible/mac/HyperTextAccessibleWrap.mm	Mon Nov 30 19:47:32 2020 +0000
+++ b/accessible/mac/HyperTextAccessibleWrap.mm	Mon Nov 30 20:46:15 2020 +0000
@@ -79,6 +79,12 @@
       mCurrentContainer = mCurrentContainer->Parent()->AsHyperText();
       mCurrentStartOffset = endOffset;
 
+      if (mCurrentContainer == mEndContainer &&
+          mCurrentStartOffset >= mEndOffset) {
+        // Reached end boundary.
+        return false;
+      }
+
       // Call NormalizeForward recursively to get top-most link if at the end of
       // one, or innermost link if at the beginning.
       NormalizeForward();
@@ -98,6 +104,12 @@
         mCurrentStartOffset = 0;
       }
 
+      if (mCurrentContainer == mEndContainer &&
+          mCurrentStartOffset >= mEndOffset) {
+        // Reached end boundary.
+        return false;
+      }
+
       // Call NormalizeForward recursively to get top-most embedding ancestor
       // if at the end of one, or innermost link if at the beginning.
       NormalizeForward();